src/stream/src/executor/sync_kv_log_store.rs

```diff
@@ -160,271 +255,252 @@ impl<S: StateStore> SyncedKvLogStoreExecutor<S> {
-    ) -> StreamExecutorResult<Option<StreamChunk>> {
-        if let Some(state_store_stream) = state_store_stream_opt {
-            match state_store_stream
-                .try_next()
-                .instrument_await("try_next item")
-                .await?
-            {
-                Some((_epoch, item)) => match item {
-                    KvLogStoreItem::StreamChunk(chunk) => Ok(Some(chunk)),
-                    KvLogStoreItem::Barrier { .. } => Ok(None),
-                },
-                None => {
-                    *state_store_stream_opt = None;
-                    Ok(None)
-                }
-            }
-        } else {
-            Ok(None)
-        }
-    }
-
-    async fn try_next_flushed_chunk_future(
-        flushed_chunk_future: &mut Option<ReadFlushedChunkFuture>,
-    ) -> StreamExecutorResult<Option<StreamChunk>> {
-        if let Some(future) = flushed_chunk_future {
-            match future.await {
-                Ok((_, chunk, _)) => {
-                    *flushed_chunk_future = None;
-                    Ok(Some(chunk))
-                }
-                Err(_) => {
-                    // TODO: log + propagate error
-                    *flushed_chunk_future = None;
-                    Ok(None)
+        metrics: &KvLogStoreMetrics,
+    ) -> StreamExecutorResult<(StreamChunk, Option<ReaderTruncationOffsetType>)> {
+        match self {
+            ReadFuture::ReadingPersistedStream(stream) => {
+                while let Some((_, item)) = stream.try_next().await? {
+                    match item {
+                        KvLogStoreItem::Barrier { .. } => {
+                            continue;
+                        }
+                        KvLogStoreItem::StreamChunk(chunk) => {
+                            // TODO: should have truncate offset when consuming historical data
+                            return Ok((chunk, None));
+                        }
+                    }
                 }
+                *self = ReadFuture::Idle;
             }
-        } else {
-            Ok(None)
+            ReadFuture::ReadingFlushedChunk { .. } | ReadFuture::Idle => {}
         }
-    }
-
-    async fn try_next_buffer_item(
-        truncation_offset: &mut Option<ReaderTruncationOffsetType>,
-        read_flushed_chunk_future: &mut Option<ReadFlushedChunkFuture>,
-        read_state: &LogStoreReadState<StateStoreFlushedSnapshotReader<S>>,
-        buffer: &mut SyncedLogStoreBuffer,
-        read_metrics: &KvLogStoreReadMetrics,
-    ) -> StreamExecutorResult<Option<StreamChunk>> {
-        let Some((item_epoch, item)) = buffer.pop_front() else {
-            return Ok(None);
-        };
-        match item {
-            LogStoreBufferItem::StreamChunk {
-                chunk, end_seq_id, ..
-            } => {
-                truncation_offset.replace((item_epoch, Some(end_seq_id)));
-                Ok(Some(chunk))
+        match self {
```

Comment:
I thought the sequence should be:
1. ReadingPersistedStream
2. Reading Flushed Chunk Future
3. Reading from buffer.

Seems like we only have 1 and 3

Reply:
Oh ic. They are independently matched. So 2&3 won't both occur.

The sequence in code is like 1, 3, 2.

Reply:
But 3 and 2 are mutually exclusive.